### ISSUE TITLE
freenas-api-nfs seems to require minimum pvc of 1GiB?

### ISSUE THREAD

**Issue by ahgraber** [democratic-csi/democratic-csi#114]
When using the new `freenas-api-nfs` driver  trying to create a PVC of size 128Mi, I get an error that the minimum size is 1GiB. This did not seem to occur using the older driver with SSH.

pvc.yaml
```
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: pvc-homer
  namespace: default
spec:
  accessModes:
    - ReadWriteOnce
  storageClassName: "csi-truenas-nfs"
  resources:
    requests:
      storage: 128Mi
```

Excerpt from `kubectl describe pvc`:

```
Warning  ProvisioningFailed    104s (x8 over 3m50s)  org.democratic-csi.truenas-nfs_k-work-03_31cd949d-72c8-43f3-aac0-cb7da97eb4a5  failed to provision volume with StorageClass "csi-truenas-nfs": rpc error: code = Internal desc = Error: {"pool_dataset_update.refquota":[{"message":"Should be greater or equal than 1073741824 or Should be 0","errno":22}]}
```

**Comment by travisghansen** [comment]
It is a new limitation actually. I asked about it but never got an answer. Let me ask again.

**Comment by travisghansen** [comment]
Still waiting on an answer for this.

In the meantime do you have more feedback (good or bad) about the api only drivers that you can share?

**Comment by ahgraber** [comment]
iSCSI driver seems solid - no issues.  I'm running into a permissions issue with nfs, but I think that may be on me; still trying to get to the root of it.

**Comment by travisghansen** [comment]
What kind of permission issue? One of the few items that changed in the config between the ssh and api variants is the chmod user/group must be specified as numbers vs names.

**Comment by ahgraber** [comment]
I thought I'd fixed that.  I resolved the errors associated with `zfs.datasetPermissionsUser` and `sfs.datasetPermissionsGroup` and the helm chart deploys successfully.

Do I also have to set `nfs.shareMaprootUser`, `nfs.shareMaprootGroup`, `nfs.shareMapallUser`, `nfs.ShareMapallGroup` with numeric codes?  (As an aside, are there best practices for setting these values?)

**Comment by travisghansen** [comment]
Entirely up to your use case honestly, but I assume most people do `root` and `wheel` (on core). I'm glad you mentioned this because although the syntax has not changed scale does not have a `wheel` group, so I do `root` and `root` for both:

```
  shareMaprootUser: root
  #shareMaprootGroup: wheel
  shareMaprootGroup: root
  shareMapallUser: ""
  shareMapallGroup: ""
```

**Comment by ahgraber** [comment]
Regardless of how I set `nfs.shareMaprootUser`, `nfs.shareMaprootGroup`, I'm still getting the following errors.  At this point, I don't _think_ it's my setup.

Current issue:  When trying to deploy a helm chart with persistence, the application pod errors out due to "access denied by server".   I can see the volume and share being created on the TrueNas side, and the PVC is created and PV bound on the k8s side.  Seems like it may be a dupe of #95, but I can't figure out where an ACL issue may exist. 

App events
```
Successfully assigned vaultwarden/vaultwarden-8446c6fdf8-vrnzx to k-work-03
---
MountVolume.MountDevice failed for volume "pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582" : rpc error: code = Internal desc = {"code":32,"stdout":"","stderr":"mount.nfs: access denied by server while mounting 10.2.1.1:/mnt/ssdpool/csi/nfs/v/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582\n","timeout":false}
---
Unable to attach or mount volumes: unmounted volumes=[vaultwarden], unattached volumes=[default-token-rm4xb vaultwarden]: timed out waiting for the condition
```

I can see that democratic-csi is creating the volume and sharing it on TrueNas.  The `csi-driver` logs are:
```
{"service":"democratic-csi","level":"info","message":"new request - driver: FreeNASApiDriver method: NodeStageVolume call: {\"_events\":{},\"_eventsCount\":1,\"call\":{},\"cancelled\":false,\"metadata\":{\"_internal_repr\":{\"user-agent\":[\"grpc-go/1.27.1\"]},\"flags\":0},\"request\":{\"publish_context\":{},\"secrets\":\"redacted\",\"volume_context\":{\"share\":\"/mnt/ssdpool/csi/nfs/v/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582\",\"storage.kubernetes.io/csiProvisionerIdentity\":\"1630980506175-8081-org.democratic-csi.truenas-nfs\",\"node_attach_driver\":\"nfs\",\"provisioner_driver\":\"freenas-api-nfs\",\"provisioner_driver_instance_id\":\"B8712266-C1B5-4DCD-9DFD-C6646C0AD3A4\",\"server\":\"10.2.1.1\"},\"volume_id\":\"pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582\",\"staging_target_path\":\"/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582/globalmount\",\"volume_capability\":{\"access_mode\":{\"mode\":\"SINGLE_NODE_WRITER\"},\"mount\":{\"mount_flags\":[\"noatime\",\"nfsvers=4\",\"sec=sys\"],\"fs_type\":\"nfs\"},\"access_type\":\"mount\"}}}"}
executing filesystem command: stat /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582/globalmount
executing mount command: findmnt --source 10.2.1.1:/mnt/ssdpool/csi/nfs/v/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582 --mountpoint /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582/globalmount --output source,target,fstype,label,options -b -J --nofsroot
executing mount command: mount -t nfs -o noatime,nfsvers=4,sec=sys,defaults 10.2.1.1:/mnt/ssdpool/csi/nfs/v/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582/globalmount
{"service":"democratic-csi","level":"error","message":"handler error - driver: FreeNASApiDriver method: NodeStageVolume error: {\"code\":32,\"stdout\":\"\",\"stderr\":\"mount.nfs: access denied by server while mounting 10.2.1.1:/mnt/ssdpool/csi/nfs/v/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582\\n\",\"timeout\":false}"}
{
code: 13,
message: '{"code":32,"stdout":"","stderr":"mount.nfs: access denied by server while mounting 10.2.1.1:/mnt/ssdpool/csi/nfs/v/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582\\n","timeout":false}'
}
{"service":"democratic-csi","level":"info","message":"new request - driver: FreeNASApiDriver method: Probe call: {\"_events\":{},\"_eventsCount\":1,\"call\":{},\"cancelled\":false,\"metadata\":{\"_internal_repr\":{\"user-agent\":[\"grpc-node/1.24.0-pre1 grpc-c/8.0.0 (linux; chttp2; game)\"]},\"flags\":0},\"request\":{}}"}
{"service":"democratic-csi","level":"info","message":"new response - driver: FreeNASApiDriver method: Probe response: {\"ready\":{\"value\":true}}"}
{"service":"democratic-csi","level":"info","message":"new request - driver: FreeNASApiDriver method: Probe call: {\"_events\":{},\"_eventsCount\":1,\"call\":{},\"cancelled\":false,\"metadata\":{\"_internal_repr\":{\"user-agent\":[\"grpc-node/1.24.0-pre1 grpc-c/8.0.0 (linux; chttp2; game)\"]},\"flags\":0},\"request\":{}}"}
{"service":"democratic-csi","level":"info","message":"new response - driver: FreeNASApiDriver method: Probe response: {\"ready\":{\"value\":true}}"}
{"service":"democratic-csi","level":"info","message":"new request - driver: FreeNASApiDriver method: NodeGetCapabilities call: {\"_events\":{},\"_eventsCount\":1,\"call\":{},\"cancelled\":false,\"metadata\":{\"_internal_repr\":{\"user-agent\":[\"grpc-go/1.27.1\"]},\"flags\":0},\"request\":{}}"}
{"service":"democratic-csi","level":"info","message":"new response - driver: FreeNASApiDriver method: NodeGetCapabilities response: {\"capabilities\":[{\"rpc\":{\"type\":\"STAGE_UNSTAGE_VOLUME\"}},{\"rpc\":{\"type\":\"GET_VOLUME_STATS\"}}]}"}
{"service":"democratic-csi","level":"info","message":"new request - driver: FreeNASApiDriver method: NodeStageVolume call: {\"_events\":{},\"_eventsCount\":1,\"call\":{},\"cancelled\":false,\"metadata\":{\"_internal_repr\":{\"user-agent\":[\"grpc-go/1.27.1\"]},\"flags\":0},\"request\":{\"publish_context\":{},\"secrets\":\"redacted\",\"volume_context\":{\"node_attach_driver\":\"nfs\",\"provisioner_driver\":\"freenas-api-nfs\",\"provisioner_driver_instance_id\":\"B8712266-C1B5-4DCD-9DFD-C6646C0AD3A4\",\"server\":\"10.2.1.1\",\"share\":\"/mnt/ssdpool/csi/nfs/v/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582\",\"storage.kubernetes.io/csiProvisionerIdentity\":\"1630980506175-8081-org.democratic-csi.truenas-nfs\"},\"volume_id\":\"pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582\",\"staging_target_path\":\"/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582/globalmount\",\"volume_capability\":{\"access_mode\":{\"mode\":\"SINGLE_NODE_WRITER\"},\"mount\":{\"mount_flags\":[\"noatime\",\"nfsvers=4\",\"sec=sys\"],\"fs_type\":\"nfs\"},\"access_type\":\"mount\"}}}"}
executing filesystem command: stat /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582/globalmount
executing mount command: findmnt --source 10.2.1.1:/mnt/ssdpool/csi/nfs/v/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582 --mountpoint /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582/globalmount --output source,target,fstype,label,options -b -J --nofsroot
executing mount command: mount -t nfs -o noatime,nfsvers=4,sec=sys,defaults 10.2.1.1:/mnt/ssdpool/csi/nfs/v/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582/globalmount
{"service":"democratic-csi","level":"error","message":"handler error - driver: FreeNASApiDriver method: NodeStageVolume error: {\"code\":32,\"stdout\":\"\",\"stderr\":\"mount.nfs: access denied by server while mounting 10.2.1.1:/mnt/ssdpool/csi/nfs/v/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582\\n\",\"timeout\":false}"}
{
code: 13,
message: '{"code":32,"stdout":"","stderr":"mount.nfs: access denied by server while mounting 10.2.1.1:/mnt/ssdpool/csi/nfs/v/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582\\n","timeout":false}'
}
```

In TrueNas, the Ganesha logs are:
```
14/09/2021 21:17:17 : epoch 612ee026 : truenas.ninerealmlabs.com : ganesha.nfsd-11463[sigmgr] vfs_create_export :FSAL :CRIT :resolve_posix_filesystem(/mnt/ssdpool/csi/nfs/v/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582) returned No such file or directory (2)
14/09/2021 21:17:17 : epoch 612ee026 : truenas.ninerealmlabs.com : ganesha.nfsd-11463[sigmgr] mdcache_fsal_create_export :FSAL :MAJ :Failed to call create_export on underlying FSAL VFS
14/09/2021 21:17:17 : epoch 612ee026 : truenas.ninerealmlabs.com : ganesha.nfsd-11463[sigmgr] fsal_cfg_commit :CONFIG :CRIT :Could not create export for (/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582) to (/mnt/ssdpool/csi/nfs/v/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582)
14/09/2021 21:17:17 : epoch 612ee026 : truenas.ninerealmlabs.com : ganesha.nfsd-11463[sigmgr] config_errs_to_log :CONFIG :CRIT :Config File (/etc/ganesha/ganesha.conf:288): 1 validation errors in block FSAL
14/09/2021 21:17:17 : epoch 612ee026 : truenas.ninerealmlabs.com : ganesha.nfsd-11463[sigmgr] config_errs_to_log :CONFIG :CRIT :Config File (/etc/ganesha/ganesha.conf:288): Errors processing block (FSAL)
14/09/2021 21:17:17 : epoch 612ee026 : truenas.ninerealmlabs.com : ganesha.nfsd-11463[sigmgr] config_errs_to_log :CONFIG :CRIT :Config File (/etc/ganesha/ganesha.conf:277): 1 validation errors in block EXPORT
14/09/2021 21:17:17 : epoch 612ee026 : truenas.ninerealmlabs.com : ganesha.nfsd-11463[sigmgr] config_errs_to_log :CONFIG :CRIT :Config File (/etc/ganesha/ganesha.conf:277): Errors processing block (EXPORT)
root@truenas:/var/log/ganesha# tail ganesha.log
14/09/2021 21:17:17 : epoch 612ee026 : truenas.ninerealmlabs.com : ganesha.nfsd-11463[sigmgr] sigmgr_thread :MAIN :EVENT :SIGHUP_HANDLER: Received SIGHUP.... initiating export list reload
14/09/2021 21:17:17 : epoch 612ee026 : truenas.ninerealmlabs.com : ganesha.nfsd-11463[sigmgr] vfs_create_export :FSAL :CRIT :resolve_posix_filesystem(/mnt/ssdpool/csi/nfs/v/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582) returned No such file or directory (2)
14/09/2021 21:17:17 : epoch 612ee026 : truenas.ninerealmlabs.com : ganesha.nfsd-11463[sigmgr] mdcache_fsal_create_export :FSAL :MAJ :Failed to call create_export on underlying FSAL VFS
14/09/2021 21:17:17 : epoch 612ee026 : truenas.ninerealmlabs.com : ganesha.nfsd-11463[sigmgr] fsal_cfg_commit :CONFIG :CRIT :Could not create export for (/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582) to (/mnt/ssdpool/csi/nfs/v/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582)
14/09/2021 21:17:17 : epoch 612ee026 : truenas.ninerealmlabs.com : ganesha.nfsd-11463[sigmgr] config_errs_to_log :CONFIG :CRIT :Config File (/etc/ganesha/ganesha.conf:288): 1 validation errors in block FSAL
14/09/2021 21:17:17 : epoch 612ee026 : truenas.ninerealmlabs.com : ganesha.nfsd-11463[sigmgr] config_errs_to_log :CONFIG :CRIT :Config File (/etc/ganesha/ganesha.conf:288): Errors processing block (FSAL)
14/09/2021 21:17:17 : epoch 612ee026 : truenas.ninerealmlabs.com : ganesha.nfsd-11463[sigmgr] config_errs_to_log :CONFIG :CRIT :Config File (/etc/ganesha/ganesha.conf:277): 1 validation errors in block EXPORT
14/09/2021 21:17:17 : epoch 612ee026 : truenas.ninerealmlabs.com : ganesha.nfsd-11463[sigmgr] config_errs_to_log :CONFIG :CRIT :Config File (/etc/ganesha/ganesha.conf:277): Errors processing block (EXPORT)
```

**Comment by travisghansen** [comment]
Can you hop on the nas and see if this dataset/folder exists?

```
/mnt/ssdpool/csi/nfs/v/pvc-97586e66-cc6d-4da5-be70-ee1b9a78c582
```

**Comment by ahgraber** [comment]
It does, and I can `cd` into it

**Comment by travisghansen** [comment]
If you restart the nfs service what do you get in logs etc?

**Comment by ahgraber** [comment]
Restarting the service worked.  Replicated teardown and deploy and still connected.  Changing the permissions (`datasetPermissionsUser` and `datasetPermissionsGroup`) required a restart of the NFS service on TrueNas.

Resolved, thanks!

**Comment by travisghansen** [comment]
Let’s make sure provisioning new volumes works without restart as well. Let’s just do a bit more testing if you can and see what else breaks.

I’m concerned that maybe something is happening asynchronous and then we’re left in a state that it doesn’t work without service restart which would not be good in production scenarios.

**Comment by ahgraber** [comment]
I've tried creating both helm deployments with persistence and manual PVCs using `freenas-api-nfs`.  Both seem to work fine with the hitch being that if I change the dataset permissions inside the `democratic-csi` deployment, I have to restart the NFS service on the NAS

**Comment by travisghansen** [comment]
OK thanks for the extra testing. That makes very little sense honestly but is helpful information before I dig in deeper and see what I observe.

Just to be clear, if you change the permissions the volume is still created successfully etc but simply mounts fail?

**Comment by ahgraber** [comment]
Yes, exactly.  The PVC is made (and bound) on k8s, the volume is created and shared on the NAS, but the pod can't mount the share because  "access denied by the server".

**Comment by travisghansen** [comment]
OK, some further info on the actual issue here:

```
historically zfs hasn't been very good with quotas (and has had nasty bugs). Specifically the logic for meeting or exceeding the quota waits for the txg to commit before returning an error. There was a fear that setting an insanely low quota expecting immediate overflow was going to cause the end-user lots of problems potentially
```

I suspect the validation for 1Gi will remain, but I will (eventually) add some logic to the csi driver to better handle this situation.

**Comment by ahgraber** [comment]
Thanks for the follow up, and sorry for double-barrel-ing the issue

**Comment by travisghansen** [comment]
More robust handling for the minimum has been added here: https://github.com/democratic-csi/democratic-csi/commit/b522a6cbb48fda5315d2b309f5cb91341903ae24

Will close after I snap a new release.

**Comment by travisghansen** [comment]
Released in `v1.4.3`.
